### PR TITLE
Fixes #19351: queryAndChekNodeId sometimes query for 0 nodes

### PR DESCRIPTION
**Pull request by ncharles**
https://issues.rudder.io/issues/19351

### CODE REVIEW

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/nodes/NodeInfoService.scala [hunk 1]
```diff
@@ -558,8 +564,13 @@ trait NodeInfoServiceCached extends NodeInfoService with NamedZioLogger with Cac
       }
     }
 
-    withUpToDateCache(s"${nodeIds.size} ldap node info") { cache =>
-      cache.values.collect{ case (x, y) if predicate(y) => x}.toSet.succeed
+    // if nodeIds is empty, return an empty set
+    if (nodeIds.nonEmpty) {
```
Comment (ncharles) [inline webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/nodes/NodeInfoService.scala:568]: this doesn"r work, as I still get report 0 ldap node info

**Comment by ncharles** [comment]
Commit modified

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/nodes/NodeInfoService.scala [hunk 2]
```diff
@@ -299,9 +299,13 @@ trait NodeInfoServiceCached extends NodeInfoService with NamedZioLogger with Cac
       def getDataIO(lastKnowModification: DateTime): IOResult[AllNodeEntries] = {
         for {
           con     <- ldap
+          t0      = System.currentTimeMillis
```
Comment (fanf) [inline webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/nodes/NodeInfoService.scala:302]: You need to use the 'current time millis' in our ZioCommon that returns an ioresult

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/nodes/NodeInfoService.scala [hunk 3]
```diff
@@ -558,8 +564,13 @@ trait NodeInfoServiceCached extends NodeInfoService with NamedZioLogger with Cac
       }
     }
 
-    withUpToDateCache(s"${nodeIds.size} ldap node info") { cache =>
-      cache.values.collect{ case (x, y) if predicate(y) => x}.toSet.succeed
+    // if nodeIds is empty, return an empty set
+    if (!nodeIds.isEmpty) {
```
Comment (ncharles) [inline webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/nodes/NodeInfoService.scala:568]: i still get 0 ldap node info

**Comment by ncharles** [comment]
PR updated with a new commit

**Comment by ncharles** [comment]
PR updated with a new commit

**Comment by ncharles** [comment]
PR updated with a new commit

**Comment by ncharles** [comment]
PR updated with a new commit

**Comment by ncharles** [comment]
PR updated with a new commit

**Comment by ncharles** [comment]
PR updated with a new commit

**Comment by ncharles** [comment]
PR rebased

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/nodes/NodeInfoService.scala [hunk 4]
```diff
@@ -398,7 +404,8 @@ trait NodeInfoServiceCached extends NodeInfoService with NamedZioLogger with Cac
     }
 
     //actual logic that check what to do (invalidate cache or not). We want it to be fully synchronized.
-    semaphore.flatMap(_.withPermit(
+    for {
```
Comment (fanf) [inline webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/nodes/NodeInfoService.scala:407]: you don't need that for anymore. It should not impact perf, but code will be clearer without it.

**Review by fanf — COMMENTED** [review]
A code-readability detail and can be mark ready to merge after that.

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/nodes/NodeInfoService.scala [hunk 5]
```diff
@@ -240,8 +240,8 @@ trait NodeInfoServiceCached extends NodeInfoService with NamedZioLogger with Cac
   def ldapMapper     : LDAPEntityMapper
   def inventoryMapper: InventoryMapper
 
-  val semaphore = Semaphore.make(1)
-
+  val semaphore = Semaphore.make(1).runNow
+  val refInt = Ref.make(0).runNow
```
Comment (fanf) [inline webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/nodes/NodeInfoService.scala:244]: you don't need that anymore

**Comment by ncharles** [comment]
PR updated with a new commit

**Review by fanf — APPROVED** [review]

**Comment by fanf** [comment]
OK, squash merging this PR

**Pull request merged**